### PR TITLE
SvelteKit: Add support for mocking `$app/state`

### PR DESCRIPTION
**Pull request by xeho91**
Resolves #31257

#### The changes in this PR are covered in the following automated tests:

- [x] stories
- [ ] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_

### Documentation

- [x] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli-storybook/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

  - `bug`: Internal changes that fixes incorrect behavior.
  - `maintenance`: User-facing maintenance tasks.
  - `dependencies`: Upgrading (sometimes downgrading) dependencies.
  - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
  - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
  - `documentation`: Documentation **only** changes. Will not show up in release changelog.
  - `feature request`: Introducing a new feature.
  - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
  - `other`: Changes that don't fit in the above categories.

   </details>

### 🦋 Canary release

This pull request has been released as version `0.0.0-pr-31369-sha-038fb8e0`. Try it out in a new sandbox by running `npx storybook@0.0.0-pr-31369-sha-038fb8e0 sandbox` or in an existing project with `npx storybook@0.0.0-pr-31369-sha-038fb8e0 upgrade`.
<details>
<summary>More information</summary>

| | |
| --- | --- |
| **Published version** | [`0.0.0-pr-31369-sha-038fb8e0`](https://npmjs.com/package/storybook/v/0.0.0-pr-31369-sha-038fb8e0) |
| **Triggered by** | @JReinhold |
| **Repository** | [xeho91/storybook](https://github.com/xeho91/storybook) |
| **Branch** | [`feat/support-mocking-sveltekit-app-state`](https://github.com/xeho91/storybook/tree/feat/support-mocking-sveltekit-app-state) |
| **Commit** | [`038fb8e0`](https://github.com/xeho91/storybook/commit/038fb8e0ce81be7af20dc5aa6f30158cbe234973) |
| **Datetime** | Mon May 19 12:21:20 UTC 2025 (`1747657280`) |
| **Workflow run** | [15112829944](https://github.com/storybookjs/storybook/actions/runs/15112829944) |

To request a new release of this pull request, mention the `@storybookjs/core` team.

_core team members can create a new canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=31369`_
</details>

## Greptile Summary

This PR adds support for mocking SvelteKit's `$app/state` module in Storybook, which is a significant enhancement for testing SvelteKit components. The changes enable developers to simulate various application states during component development and testing.

The implementation includes:
- A new mock implementation for the `$app/state` module with support for page, navigating, and updated states
- Updated type definitions and parameter structures
- Migration from filesystem paths to package imports for better maintainability
- Comprehensive example stories demonstrating the mocking functionality
- Updated documentation reflecting the new features

The changes also maintain backward compatibility by supporting both the new state-based approach and the legacy stores approach.

## Confidence score: 4/5

1. This PR is generally safe to merge as it adds new functionality without breaking existing features
2. The score of 4 reflects strong implementation with proper types, documentation, and example stories, though lacking unit tests
3. Key files needing attention:
   - code/frameworks/sveltekit/src/mocks/app/state.svelte.js for potential edge cases
   - code/frameworks/sveltekit/src/preview.ts for state initialization logic
   - code/frameworks/sveltekit/src/types.ts for type safety

<sub>10 files reviewed, 1 comment</sub>
<sub>[Edit PR Review Bot Settings](https://app.greptile.com/review/github) | [Greptile](https://greptile.com?utm_source=greptile_expert&utm_medium=github&utm_campaign=code_reviews&utm_content=storybook_31369)</sub>

## Summary by CodeRabbit

- **New Features**
  - Experimental $app/state mocking for SvelteKit (page, navigating, updated) with module alias so $app/state resolves to the mock.
  - Automatic state sync per story via a new preview hook.
  - Example State component and stories demonstrating Page, Navigating, and Updated scenarios.
  - Public types/parameters added to support sveltekit_experimental.state and richer state shapes.

- **Documentation**
  - SvelteKit guide updated with $app/state API, examples, and options; stores-based docs retained.

- **Refactor**
  - Renamed store setter helpers to setAppStoresPage / setAppStoresNavigating / setAppStoresUpdated.

- **Chores**
  - Added packaging/build export and dependency optimization entries to surface the static state mock.

### CODE REVIEW

File: code/frameworks/sveltekit/src/mocks/app/state.svelte.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+};
+
+let updatedCurrent = $state.raw(false);
+
+export let updated = {
+  current: updatedCurrent,
+  check: () =>
+    new Promise<boolean>((resolve) => {
+      resolve(updatedCurrent);
+    }),
```
[the remark this answers is not in the document]
Comment (JReinhold) [reply code/frameworks/sveltekit/src/mocks/app/state.svelte.ts:59]: this, but also wrap it in `fn()` from `storybook/test`, so that it shows up in the actions panel and can be changed when defining stories:

```js
import { fn } from 'storybook/test';
...
let updatedCurrent = $state.raw(false);

export let updated = {
  current: updatedCurrent,
  check: fn(() => Promise.resolve(updatedCurrent)),
};
```

**Review by JReinhold — COMMENTED** [review]
I think the greptile comments are sound.

Also, remember to update the docs:
- the table https://github.com/storybookjs/storybook/blob/next/docs/get-started/frameworks/sveltekit.mdx/#L81
- The API docs https://github.com/storybookjs/storybook/blob/next/docs/get-started/frameworks/sveltekit.mdx/#L342

File: code/frameworks/sveltekit/src/preview.ts [hunk 2]
```diff
@@ -21,6 +26,32 @@ const svelteKitMocksDecorator: Decorator = (Story, ctx) => {
...
+  assignStateProperties(svelteKitParameters.state?.navigating, appStateNavigating, [
+    'from',
+    'to',
+    'type',
+    'willUnload',
+    'delta',
+    'complete',
+  ]);
+
+  console.log({ svelteKitParameters });
```
Comment (JReinhold) [inline code/frameworks/sveltekit/src/preview.ts:49]:
```suggestion
```

File: docs/get-started/frameworks/sveltekit.mdx [hunk 3]
```diff
@@ -351,6 +352,30 @@ Type: boolean
...
+
+Provides mocks for the [`$app/state`](https://svelte.dev/docs/kit/$app-state) module.
+
+##### `state.navigating`
+
+Type: See [SvelteKit docs](https://svelte.dev/docs/kit/$app-state#navigating)
+
+A partial version of the [`navigating`](https://svelte.dev/docs/kit/$app-state#navigating) store.
+
+##### `stores.page`
```
Comment (JReinhold) [inline docs/get-started/frameworks/sveltekit.mdx:367]:
```suggestion
##### `state.page`
```

File: docs/get-started/frameworks/sveltekit.mdx [hunk 4]
```diff
@@ -351,6 +352,30 @@ Type: boolean
...
+
+Type: See [SvelteKit docs](https://svelte.dev/docs/kit/$app-state#page)
+
+A partial version of the [`page`](https://svelte.dev/docs/kit/$app-state#page) store.
+
+##### `stores.updated`
```
Comment (JReinhold) [inline docs/get-started/frameworks/sveltekit.mdx:373]:
```suggestion
##### `state.updated`
```

File: code/frameworks/sveltekit/src/mocks/app/state.svelte.ts [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+  params: Params;
+  route: {
+    id: RouteId;
+  };
+  status: number;
+  error: App.Error | null;
+  data: App.PageData & Record<string, any>;
+  state: App.PageState;
+  form: any;
+}
```
Comment (JReinhold) [inline code/frameworks/sveltekit/src/mocks/app/state.svelte.ts:55]:
```suggestion
```

Given that no-one imports from this file directly - it's always used as an alias - I don't think there's any value to exporting the types.

File: code/frameworks/sveltekit/src/types.ts [hunk 6]
```diff
@@ -7,6 +7,7 @@ import type { BuilderOptions, StorybookConfigVite } from '@storybook/builder-vit
 
 import type { enhance } from './mocks/app/forms';
 import type { goto, invalidate, invalidateAll } from './mocks/app/navigation';
+import type { Navigation, Page } from './mocks/app/state.svelte';
```
Comment (JReinhold) [inline code/frameworks/sveltekit/src/types.ts:10]: this might be the reason for the failure. We should try defining the types in this file directly, so we don't have to import the `.svelte.ts` file.

**Review by JReinhold — CHANGES_REQUESTED** [review]

File: code/frameworks/sveltekit/src/mocks/app/state.svelte.js [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+export let page = {
+  data: pageData,
+  form: pageForm,
+  error: pageError,
+  params: pageParams,
+  route: pageRoute,
+  state: pageState,
+  status: pageStatus,
+  url: pageUrl,
+};
```
Comment (paoloricciuti) [inline code/frameworks/sveltekit/src/mocks/app/state.svelte.js:64]: If you want this to be updated it can't just have a reference to the raw state...all of this needs to be getters/setters
```suggestion
export let page = {
  get data() {
    return pageData;
  },
  set data(newPageData: typeof pageData) {
    pageData = newPageData;
  },
  get form() {
    return pageForm;
  },
  set form(newPageForm: typeof pageForm) {
    pageForm = newPageForm;
  },
  get error() {
    return pageError;
  },
  set error(newPageError: typeof pageError) {
    pageError = newPageError;
  },
  get params() {
    return pageParams;
  },
  set params(newPageParams: typeof pageParams) {
    pageParams = newPageParams;
  },
  get route() {
    return pageRoute;
  },
  set route(newPageRoute: typeof pageRoute) {
    pageRoute = newPageRoute;
  },
  get state() {
    return pageState;
  },
  set state(newPageState: typeof pageState) {
    pageState = newPageState;
  },
  get status() {
    return pageStatus;
  },
  set status(newPageStatus: typeof pageStatus) {
    pageStatus = newPageStatus;
  },
  get url() {
    return pageUrl;
  },
  set url(newPageUrl: typeof pageUrl) {
    pageUrl = newPageUrl;
  },
};
```

File: code/frameworks/sveltekit/src/mocks/app/state.svelte.js [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+/** @type {Navigation} */
+export let navigating = {
+  from: navigatingFrom,
+  to: navigatingTo,
+  type: navigatingType,
+  willUnload: navigatingWillUnload,
+  delta: navigatingDelta,
+  complete: navigatingComplete,
+};
```
Comment (paoloricciuti) [inline code/frameworks/sveltekit/src/mocks/app/state.svelte.js:107]: Same thing here
```suggestion
export let navigating = {
  get from() {
    return navigatingFrom;
  },
  set from(newNavigatingFrom: typeof navigatingFrom) {
    navigatingFrom = newNavigatingFrom;
  },
  get to() {
    return navigatingTo;
  },
  set to(newNavigatingTo: typeof navigatingTo) {
    navigatingTo = newNavigatingTo;
  },
  get type() {
    return navigatingType;
  },
  set type(newNavigatingType: typeof navigatingType) {
    navigatingType = newNavigatingType;
  },
  get willUnload() {
    return navigatingWillUnload;
  },
  set willUnload(newNavigatingWillUnload: typeof navigatingWillUnload) {
    navigatingWillUnload = newNavigatingWillUnload;
  },
  get delta() {
    return navigatingDelta;
  },
  set delta(newNavigatingDelta: typeof navigatingDelta) {
    navigatingDelta = newNavigatingDelta;
  },
  get complete() {
    return navigatingComplete;
  },
  set complete(newNavigatingComplete: typeof navigatingComplete) {
    navigatingComplete = newNavigatingComplete;
  },
};
```

**Comment by paoloricciuti** [comment]
The suggestion I added should be enough to fix the reactivity problem.

**Comment by JReinhold** [comment]
@paoloricciuti always coming in clutch. 😍

File: code/frameworks/sveltekit/src/mocks/app/state.svelte.js [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+/** @type {Navigation['to'] | null} */
+let navigatingTo = $state.raw(defaultStatePageValues.to);
+/** @type {Navigation['type'] | null} */
+let navigatingType = $state.raw(defaultStatePageValues.type);
+/** @type {Navigation['willUnload'] | null} */
+let navigatingWillUnload = $state.raw(defaultStatePageValues.willUnload);
+/** @type {Navigation['delta'] | null} */
+let navigatingDelta = $state.raw(defaultStatePageValues.delta);
+/** @type {Navigation['complete'] | null} */
+let navigatingComplete = $state.raw(defaultStatePageValues.complete);
```
[the remark this answers is not in the document]
Comment (xeho91) [reply code/frameworks/sveltekit/src/mocks/app/state.svelte.js]: Oh, thanks!

File: code/frameworks/sveltekit/src/preview.ts [hunk 10]
```diff
@@ -136,3 +148,11 @@ const svelteKitMocksDecorator: Decorator = (Story, ctx) => {
...
 
 export const decorators: Decorator[] = [svelteKitMocksDecorator];
+
+export const beforeEach: Preview['beforeEach'] = async (ctx) => {
+  const svelteKitParameters: SvelteKitParameters = ctx.parameters?.sveltekit_experimental ?? {};
+
+  setAppStatePage(svelteKitParameters?.state?.page);
+  setAppStateNavigating(svelteKitParameters?.state?.navigating);
+  setAppStateUpdated(svelteKitParameters?.state?.updated);
+};
```
Comment (JReinhold) [inline code/frameworks/sveltekit/src/preview.ts:158]: @xeho91 so presumably the failing tests indicate that we need to move this into the decorator instead?

Comment (xeho91) [reply code/frameworks/sveltekit/src/preview.ts:158]: Not entirely sure bit I think no. I'm puzzled by this error. The mocks for `$app/state` doesn't use `getContext` at all.

The mocks for `$app/stores` do use the `getContext`, and they remain in the decorator (I only improved for set functions).

**Review by OrangeDoro — COMMENTED** [review]
Hi! I'm a grad student working on a research project about using large language models to automate code review. Based on your commit 038fb8e0ce81be7af20dc5aa6f30158cbe234973 and the changes in code/frameworks/sveltekit/src/mocks/app/state.svelte.js, my tool generated this comment:
1. **Null Handling**: Ensure that the subsequent code that uses these variables properly handles `null` values to avoid null dereference errors.
2. **Data Integrity**: Ensure that `defaultStateNavigatingValues` is properly validated before being used.
3. **Type Safety**: Ensure that the types being assigned from `defaultStateNavigatingValues` are compatible with the expected types in the `Navigation` type definition.
4. **Consistency of State Values**: Ensure that `defaultStateNavigatingValues` contains the correct and intended values for `from`, `to`, `type`, `willUnload`, `delta`, and `complete`.
5. **State Management**: Ensure that `$state.raw()` does not expose sensitive information or allow for state manipulation that could lead to security issues.
6. **Avoiding Repetition**: The repeated pattern of using `$state.raw(defaultStateNavigatingValues.<property>)` can be abstracted into a function or a mapping object to reduce redundancy.
7. **Potential for Future State Management**: Consider implementing a more dynamic state management approach to allow for easier addition of new navigation states.
8. **Testing**: Run tests that cover the navigation functionality to ensure that the application behaves as expected.
9. **Documentation**: Update any comments or documentation that refer to `defaultStatePageValues` to reflect the new source of truth (`defaultStateNavigatingValues`).

As part of my research, I'm trying to understand how useful these comments are in real-world development. If you have a moment, I'd be super grateful if you could quickly reply to these two yes/no questions:


1. Does this comment provide suggestions from a dimension you hadn’t considered?

2. Do you find this comment helpful?

Thanks a lot for your time and feedback! And sorry again if this message is a bother.

**Review by JReinhold — APPROVED** [review]
🎉
